### PR TITLE
feat(medusa-react,medusa-js): Allow custom headers

### PR DESCRIPTION
**Pull request by pevey**
Adds support for passing custom headers to backend through medusa-js and medusa-react.

This is something I needed to be able to easily access the server behind Cloudflare Access using service tokens, so I thought I would propose it upstream.  It can be helpful in a number of deployment situations to be able to pass custom headers in requests to the backend.

### CODE REVIEW

**Review by olivermrbl — APPROVED** [review]
think this makes sense - thank you for the contribution!

File: packages/medusa-js/src/request.ts [hunk 1]
```diff
@@ -199,6 +200,9 @@ class Client {
     options: RequestOptions = {},
     customHeaders: Record<string, any> = {}
   ): Promise<any> {
+   
+    customHeaders = { ...customHeaders, ...this.config.customHeaders }
```
Comment (adrien2p) [inline packages/medusa-js/src/request.ts]: suggestion: I think that we should invert the sense here, the custom headers from the config first and if some custom headers are provided when calling it then it should take priority. wdyt?

**Review by adrien2p — APPROVED** [review]
LGTM with comment

File: packages/medusa-js/src/request.ts (hunk 1 above)
Comment (olivermrbl) [reply packages/medusa-js/src/request.ts]: Good point. Think it makes sense to allow developers to overwrite custom headers on a per-request basis.

Comment (pevey) [reply packages/medusa-js/src/request.ts]: The intention is to allow override, but the order needs to be what was initially proposed for an override to work.  The second element overrides the first.

Comment (olivermrbl) [reply packages/medusa-js/src/request.ts]: The idea with the `customHeaders` configuration is global headers - headers shared among all requests - right? Wouldn't allowing overrides on a per-request basis make sense, which I think is what Adrien proposes? 

> the order needs to be what was initially proposed for an override to work.

Update: I may be misinterpreting this comment. Can you elaborate?  ^

Comment (adrien2p) [reply packages/medusa-js/src/request.ts]: Yes, this is what I proposed, so the config headers are applied but overridden by the request custom headers. With the current approach, if someone wants to change the custom headers from the config they won't be able to as they are applied at last

Comment (pevey) [reply packages/medusa-js/src/request.ts]: Either way is fine for my use case.  The headers I'll be using won't conflict.  I defer to you guys on what should override.

**Review by olivermrbl — APPROVED** [review]
LGTM!

**Pull request merged**